### DISCUSSION TITLE
Inserts & updates without notifying the stream manager.

### DISCUSSION THREAD

**Discussion by p-kager** [simolus3/drift#2104] [Q&A] [answered]
Hi, I probably have the answer but just making sure in which case this would be more of an enhancement suggestion. I am running drift (Flutter) in an isolate where then another isolate is doing a bit of a heavy lifting. It's running a bunch of transactions that include updates, inserts and even selects. At the end of each transaction as expected the stream manager gets notified and the streamed queries used by the UI get a re-run. What I'd like to be able to do at the start of the transaction is to disable the stream manager notification. The reason is that I do not want to be running nested transaction but rather one after another and only then notify the stream manager. The reason is that the database gets locked and I would like to leave gaps for the main isolate to access it as well (accessing partial but consistent data).
From what I could work out is to run custom queries inside the transactions and then calling markTablesUpdated(). This creates a little bit of inconvenience by having to write the queries but it is a way I can solve this.

Any input would be appreciated.

P.S - Really like the drift package. Some excellent work went into it.

**Comment by simolus3** [comment, marked as answer]
Do you need stream query synchronization between the two isolates at all? If you're not using a `DriftIsolate`, updates on one isolate won't affect streams in the other. Drift implements a notification framework across isolates for this to work. If you just independenty open two databases on two isolates without a `DriftIsolate` for communication, stream queries are entirely local.

If you want to use a drift isolate, but without stream queries, construct a `DatabaseConnection`  with the original executor you got from `DriftIsolate.connect` but replace ths stream queries with a new `StreamQueryStore` which will be entirely local then.

**Reply by p-kager** [reply]
Hi Simon,

Thanks for the suggestion. I've had a look at bypassing the ```StreamQueryStore``` like so:

```
DatabaseConnection conn = DatabaseConnection(
  (await DriftIsolate.fromConnectPort(driftDbSendPort).connect()).executor,
);
AppDatabase db = AppDatabase.connect(conn);
```

This would prevent query streams on the UI isolate being updated even after calling ```markTablesUpdated()``` which is what I expected, but not quite solving my problem. 
I think I could also keep two database class instances within the worker isolate, one that would be used to update the streams only, by calling ```markTablesUpdated()``` as and when needed, and one with local ```StreamQueryStore``` to do the inserts and updates. 

Or I could be using custom queries that do not automatically update the query streams. 

I am not sure how much work would be involved in having something like ```db.transaction(() {}, updateStreams: false)``` and also not sure if I would be the only one finding it useful, in which case I should probably reconsider my whole approach :sweat_smile:.

**Reply by simolus3** [reply]
> and also not sure if I would be the only one finding it useful, in which case I should probably reconsider my whole approach :sweat_smile:.

The main problem here are the frequent, small transactions that cause streams in the UI isolate to update, correct?

Inside a transaction, drift collects updates and will dispatch them at once after the transaction completes. So a single transaction will only cause a single stream update for streams created outside of that transaction.

So if the small transactions are part of a larger control flow instead of independent events, you could wrap all of them in a single `transaction` block. Drift supports nested transactions in recent versions, so you can just wrap them in a larger transaction and use that. Only a single update should be forwarded to peers then.

Having an `updateStreams: false` API seems kind of hacky to me. There should be no use-case where that is a good option ;D If you need to batch up stream updates _dynamically_, you could write a custom class for that:

```dart
/// A stream query store implementation that works like a regular store for
/// local streams, but will only forward stream updates to a remote store
/// if not paused.
class PausingStreamUpdates extends StreamQueryStore {
  final StreamQueryStore _inner;

  late final StreamSubscription<Set<TableUpdate>> _updatesFromRemote;

  // A set of updates to batch up while paused. Null iff not paused.
  Set<TableUpdate>? _pendingUpdates;
  bool _currentlyForwardingUpdates = false;

  PausingStreamUpdates(this._inner) {
    // Directly forward table updates received from the remote isolate to the
    // local queries.
    _updatesFromRemote = _inner
        .updatesForSync(TableUpdateQuery.any())
        .where((_) => !_currentlyForwardingUpdates)
        .listen(
          super.handleTableUpdates,
        );
  }

  @override
  void handleTableUpdates(Set<TableUpdate> updates) {
    // Handle local streams
    super.handleTableUpdates(updates);

    if (_pendingUpdates != null) {
      _pendingUpdates!.addAll(updates);
    } else {
      // Not paused, forward to remotes
      _currentlyForwardingUpdates = true;
      _inner.handleTableUpdates(updates);
      _currentlyForwardingUpdates = false;
    }
  }

  void pause() {
    _pendingUpdates ??= {};
  }

  void resume() {
    final pending = _pendingUpdates;
    _pendingUpdates = null;

    // Notify remotes now.
    if (pending != null && pending.isNotEmpty) {
      _inner.handleTableUpdates(pending);
    }
  }

  @override
  Future<void> close() async {
    await _updatesFromRemote.cancel();
    return super.close();
  }
}
```

I didn't actually test this (:crossed_fingers:) and it still feels pretty hacky. But something like this should work to dynamically pause stream updates from a local database to other isolate. If it doesn't work, let me know and I'll put in more effort to get it working. In the worker isolate (the one causing all the updates), connect like this:

```dart
final original = await DriftIsolate.fromConnectPort(driftDbSendPort).connect();
DatabaseConnection conn = DatabaseConnection(original.executor, streamQueries: PausingStreamUpdates(original.streamQueries));
```

Then, you use `pause()` to stop the worker isolate from reporting updates to the UI and `resume()` to atomically notify the UI of all updates from the pause.

**Reply by p-kager** [reply]
It really comes down to locking of the database for too long by doing a lot of work within a transaction. Every insert and update on one table (arbitrary number, pushed by server sent event in batches) updates a field in another table. This field can also be updated by the UI thread by user interaction hence the need for a transaction. When the background transaction is executing the UI needs to wait for the database lock to release and I'd like to limit that to as little time as possible, by splitting the background work into multiple transactions. Each one of these transactions updates the query streams, but in my case it is not always desirable. Depending on the data sent by the server, I need to notify the streams after:

- every transaction (which works by using the normal functionality)
- after a full batch that was split into transactions

I do agree that in a common scenario updating the stream after a transaction is complete is pretty much always desirable and pausing the stream feels hacky, but I am not sure I can get around it in my case.

Anyways, thanks a lot for your help and the code above. It looks like what I need so I'll give it a shot. :+1:
